From 650c53b3e0c69bbe81ccda4b7dbd1eed75adcefd Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Thu, 30 Jan 2020 16:50:08 +0800
Subject: [PATCH 065/106] bmq: Introduce boot-time kernel parameter
 "bmq.timeslice".

Introduce boot-time kernel parameter "bmq.timeslice" in us. Remove
kernel config CONFIG_SCHED_TIMESLICE.
---
 .../admin-guide/kernel-parameters.txt         |  5 ++++
 init/Kconfig                                  |  8 ------
 kernel/sched/bmq.c                            | 27 ++++++++++++++-----
 3 files changed, 25 insertions(+), 15 deletions(-)

diff --git a/Documentation/admin-guide/kernel-parameters.txt b/Documentation/admin-guide/kernel-parameters.txt
index 7bc83f3d9bdf..e549b4d53746 100644
--- a/Documentation/admin-guide/kernel-parameters.txt
+++ b/Documentation/admin-guide/kernel-parameters.txt
@@ -438,6 +438,11 @@
 			embedded devices based on command line input.
 			See Documentation/block/cmdline-partition.rst
 
+	bmq.timeslice=	[KNL] Time slice in us for BMQ scheduler.
+			Format: <int> (must be >= 1000)
+			Default: 4000
+			See Documentation/scheduler/sched-BMQ.txt
+
 	boot_delay=	Milliseconds to delay each printk during boot.
 			Values larger than 10 seconds (10000) are changed to
 			no delay (0).
diff --git a/init/Kconfig b/init/Kconfig
index 6f47609e6b61..850f730faef5 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -699,14 +699,6 @@ config SCHED_BMQ
           Say Y here.
 	default y
 
-config SCHED_TIMESLICE
-	int "Scheduler Task time slice"
-	depends on SCHED_BMQ
-	help
-	  Time slice in ms for BMQ CPU scheduler, default 4 ms.
-	default 2 if PREEMPT
-	default 4 if !PREEMPT
-
 config UCLAMP_TASK
 	bool "Enable utilization clamping for RT/FAIR tasks"
 	depends on CPU_FREQ_GOV_SCHEDUTIL
diff --git a/kernel/sched/bmq.c b/kernel/sched/bmq.c
index 189865fa0597..5b59f735f662 100644
--- a/kernel/sched/bmq.c
+++ b/kernel/sched/bmq.c
@@ -50,7 +50,20 @@
 
 #define STOP_PRIO		(MAX_RT_PRIO - 1)
 
-#define SCHED_TIMESLICE_NS	(CONFIG_SCHED_TIMESLICE * 1000 * 1000)
+/* Default time slice is 4 in ms, can be set via kernel parameter "bmq.timeslice" */
+u64 sched_timeslice_ns __read_mostly = (4 * 1000 * 1000);
+
+static int __init sched_timeslice(char *str)
+{
+	int timeslice_us;
+
+	get_option(&str, &timeslice_us);
+	if (timeslice_us >= 1000)
+		sched_timeslice_ns = timeslice_us * 1000;
+
+	return 0;
+}
+early_param("bmq.timeslice", sched_timeslice);
 
 /* Reschedule if less than this many Î¼s left */
 #define RESCHED_NS		(100 * 1000)
@@ -81,7 +94,7 @@ static inline void print_scheduler_version(void)
 int sched_yield_type __read_mostly = 1;
 
 #define rq_switch_time(rq)	((rq)->clock - (rq)->last_ts_switch)
-#define boost_threshold(p)	(SCHED_TIMESLICE_NS >>\
+#define boost_threshold(p)	(sched_timeslice_ns >>\
 				 (10 - MAX_PRIORITY_ADJ -  (p)->boost_prio))
 
 static inline void boost_task(struct task_struct *p)
@@ -1655,7 +1668,7 @@ static int try_to_wake_up(struct task_struct *p, unsigned int state,
 		atomic_dec(&task_rq(p)->nr_iowait);
 	}
 
-	if(cpu_rq(smp_processor_id())->clock - p->last_ran > SCHED_TIMESLICE_NS)
+	if(cpu_rq(smp_processor_id())->clock - p->last_ran > sched_timeslice_ns)
 		boost_task(p);
 
 	cpu = select_task_rq(p);
@@ -1776,7 +1789,7 @@ int sched_fork(unsigned long __maybe_unused clone_flags, struct task_struct *p)
 #endif
 
 	if (p->time_slice < RESCHED_NS) {
-		p->time_slice = SCHED_TIMESLICE_NS;
+		p->time_slice = sched_timeslice_ns;
 		resched_curr(rq);
 	}
 	raw_spin_unlock_irqrestore(&rq->lock, flags);
@@ -2800,7 +2813,7 @@ static inline void check_curr(struct task_struct *p, struct rq *rq)
 	update_curr(rq, p);
 
 	if (p->time_slice < RESCHED_NS) {
-		p->time_slice = SCHED_TIMESLICE_NS;
+		p->time_slice = sched_timeslice_ns;
 		if (SCHED_FIFO != p->policy && task_on_rq_queued(p)) {
 			if (SCHED_RR != p->policy)
 				deboost_task(p);
@@ -2975,7 +2988,7 @@ static inline void set_rq_task(struct rq *rq, struct task_struct *p)
 {
 	p->last_ran = rq->clock_task;
 
-	if (unlikely(SCHED_TIMESLICE_NS == p->time_slice))
+	if (unlikely(sched_timeslice_ns == p->time_slice))
 		rq->last_ts_switch = rq->clock;
 #ifdef CONFIG_HIGH_RES_TIMERS
 	if (p != rq->idle)
@@ -4789,7 +4802,7 @@ static int sched_rr_get_interval(pid_t pid, struct timespec64 *t)
 		goto out_unlock;
 	rcu_read_unlock();
 
-	*t = ns_to_timespec64(SCHED_TIMESLICE_NS);
+	*t = ns_to_timespec64(sched_timeslice_ns);
 	return 0;
 
 out_unlock:
-- 
2.27.0.rc2

